### ISSUE TITLE
Spurious type error using result of re.match in lambda inside map

### ISSUE THREAD

**Issue by alexcoplan** [python/mypy#18947]
Consider the following testcase:

```python
import re
r = re.compile("([a-z]+)")
if m := re.match(r, "abc"):
  matches = map(lambda i: m.group(i), [1])
  print(next(matches))
```

I expect this to pass type checking successfully, but mypy instead complains about the map line as follows:

```
mypy ./t.py
t.py:4: error: Item "None" of "Match[str] | None" has no attribute "group"  [union-attr]
Found 1 error in 1 file (checked 1 source file)
```

The bug only seems to happen if the lambda appears directly inside the call to map. E.g. if I pre-declare it as follows:

```python
import re
r = re.compile("([a-z]+)")
if m := re.match(r, "abc"):
  fn = lambda i: m.group(i)
  matches = map(fn, [1])
  print(next(matches))
```

then this type-checks OK.  My environment is as follows:

```
$ mypy --version
mypy 1.15.0 (compiled: yes)
$ python3 --version
Python 3.13.2
```

**Comment by A5rocks** [comment]
Hmm my first instinct is that this is intentional because mypy tries not to persist type narrowing into callbacks (which it cannot track the lifetimes of). But moving the lambda out wouldn't fix this in that case...

**Comment by sterliakov** [comment]
Yes, what Eric mentions above is indeed a bug, `lambda` return expr is handled in a wrong scope. 

```python
x: int | None
if x is not None:
    bad = lambda: reveal_type(x)  # N: Revealed type is "builtins.int"
    def bad_fn():
        reveal_type(x)  # N: Revealed type is "Union[builtins.int, None]"
        return -x  # E: Unsupported operand type for unary - ("int | None")  [operator]
```

Currently `def` functions use a new instance of binder and copy everything "safe" from global scope into it, and lambdas do not undergo similar treatment. This is a fairly difficult problem to fix.
